Refactor Nix code

toastal opened the pull request:
Y’all got close on the refactor, but missed some important things:

* use overlays
* expose those overlays
* make sure `callPackage` is used with actual derivations (not attrsets)
* expose those packages individually for downstream
* filter sources to avoid unnecessary rebuilds

This works now using `nix-build release.nix -A package` (or `nix-build release.nix` to build all packages)… works with & without flakes. Can still be updated with `nix flake update nixpkgs`.

(also I am looking for Nix work if y’all know anyone looking for that, even short-term & contracts)

toastal committed:
nix: make alice package

toastal committed:
nix: make default overlay

toastal committed:
nix: make release.nix

toastal committed:
nix: make dev-shell

toastal committed:
nix: make development overlay

toastal committed:
nix: expose overlays in flake

toastal committed:
nix: make tools

toastal committed:
nix: make versioned overlay

toastal committed:
nix: use default from release.nix

toastal committed:
nix: redo flake to use overlays + nixpkgs.lib

toastal committed:
nix: alice’s mainProgram

gridbugs commented:
Thanks for the patch! I think it's a little complicated for my needs, and I don think I'll be able to maintain the nix code in this repo if I was to merge this PR due to my limited understanding of nix. Still I think there are some simple improvements that you've made here that I can learn from when writing nix code in the future.

Good luck with your job search!

toastal commented:
It ends up being about as simple as it can be. You can also [reach out](https://toast.al/contact/) about anything in particular that isn’t understood as it’s about as simple as it gets but also remaining flexible for all use cases.

toastal committed:
nix: move climate to overlay

toastal committed:
nix: dev-shell from packages

toastal committed:
nix: expose versioned

toastal committed:
nix: expose tools

gridbugs committed:
Rename top level namespace to alicecaml

This avoids a conflict with the package
https://github.com/cyrealtype/Alice already in the nixpkgs repo.

gridbugs committed:
Fix overriding alice version

gridbugs committed:
Correct version in derivation name when overriding version

Previously the derivation name would always include the current
development version rather than than the installed version when
overriding the version.

gridbugs commented:
Rebased and fixed a couple of issues I ran into while playing around with this.

gridbugs committed:
Consistent naming of final ofinal and oprev

gridbugs committed:
Don't build graphviz with muslc in dev shell

This change gives the dev shell control over which dependencies are from
pkgsMusl and which are from the native package repo. Some dependencies
(currently just graphviz) don't need to be built with muslc, and building
them with muslc takes a long time.

gridbugs committed:
Add back derivations with alice and tools

This restores functionality where Alice can be installed alongside the
OCaml tools as a single derivation, and makes the derivation the
default.

gridbugs committed:
Remove __functor from default.nix

I don't know how to maintain this field in the face of future changes to
the nix files.

gridbugs committed:
More verbose names to avoid confusion about tools

gridbugs committed:
Update dev flake name in test

toastal commented:
oh neat
